### DISCUSSION TITLE
What is the proper way of getting Content Types of a specific Group

### DISCUSSION THREAD

**Discussion by plamber** [pnp/pnpcore#615] [Q&A] [answered]
Hello,
I was wondering what is the most efficient way to retrieve all content types assigned to a specific Group without requiring to retrieve all content types first and then perform a in-memory search?

Thank you for your feedback,
Patrick

**Comment by jansenbe** [comment, marked as answer]
Hi @plamber ,

You can use any of the below approaches:

```csharp
foreach(var contentType in context.Web.ContentTypes.Where(p => p.Group == "List Content Types"))
{
}

// or

await foreach(var contentType in context.Web.ContentTypes.Where(p => p.Group == "List Content Types").AsAsyncEnumerable())
{
}

// or 

var contentTypes = await context.Web.ContentTypes.Where(p => p.Group == "List Content Types").ToListAsync();

// or

var contentTypes = context.Web.ContentTypes.Where(p => p.Group == "List Content Types").ToList()
```

**Reply by plamber** [reply]
Thank you. So simple

I think I have to go back to the basics of the SDK to understand better how it behaves.

I still have issues understanding when I am triggering a call to the service and when I am working in memory.

Is there a way for me to see which calls are performed in the background while debugging in visual studio using the nuget package?

**Reply by jansenbe** [reply]
@plamber : https://pnp.github.io/pnpcore/using-the-sdk/basics-getdata.html describes the general data retrieval. When getting data for a model (Web, List,...) the `Load...` methods load data into the in-memory model (so the context you're using), using `Get...` methods loads the data is the output variable. When getting data for collections (e.g. lists, list items,...) similar logic applies, meaning doing a `context.Web.Load(p => p.Lists)` will populate the collection in memory, doing a `var lists = context.Web.Get(p=>p.Lists)` gets you an `IWeb` variable with the `Lists` collection populated. 

Once you have data in memory (either under `context.Web.something` or under `variable.Web.something`) querying that data is best done by using the `AsRequested()` method. Since we have our own LINQ provider doing for example a `.FirstOrDefault()` on `context.Web.Lists` would otherwise result in a new query to the SharePoint. 

There a number of options to see when an actual call is being made, one is setting a breakpoint in the `BatchClient` on method `internal async Task<List<BatchResult>> ExecuteBatch(Batch batch)`, but what I usually do is check log output. If you put `Debug` logging on each query will we written out in the logs. https://github.com/pnp/pnpcore/blob/dev/samples/Demo.Blazor/wwwroot/appsettings.json is sample config file showing that, here's the docs: https://pnp.github.io/pnpcore/using-the-sdk/basics-settings.html.

I'm more than happy to further clarify things in documentation, feel free to propose suggestions.

**Reply by plamber** [reply]
Hi,
Thank you for the clarification.

This will definitely help.

Cheers
